src/lightning/data/streaming/resolver.py

```diff
@@ -0,0 +1,340 @@
+import datetime
+import os
+import re
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from time import sleep
+from typing import Optional, Union
+from urllib import parse
+
+from lightning_cloud.openapi import V1CloudSpace
+from lightning_cloud.rest_client import LightningClient
+
+# To avoid adding lightning_utilities as a dependency for now.
+try:
+    import boto3
+    import botocore
+
+    _BOTO3_AVAILABLE = True
+except Exception:
+    _BOTO3_AVAILABLE = False
+
+
+try:
+    from lightning_sdk import Machine, Studio
+
+    _LIGHTNING_SDK_AVAILABLE = True
+except (ImportError, ModuleNotFoundError):
+
+    class Machine:  # type: ignore
+        pass
+
+    _LIGHTNING_SDK_AVAILABLE = False
+
+
+@dataclass
+class Dir:
+    """Holds a directory path and possibly its associated remote URL."""
+
+    path: Optional[str] = None
+    url: Optional[str] = None
+
+
+def _resolve_dir(dir_path: Optional[Union[str, Dir]]) -> Dir:
+    if isinstance(dir_path, Dir):
+        return Dir(path=str(dir_path.path) if dir_path.path else None, url=str(dir_path.url) if dir_path.url else None)
+
+    if dir_path is None:
+        return Dir()
+
+    if not isinstance(dir_path, str):
```

Comment:
what about pathlib.Path here?

Reply:
This is provided by the user and it is most likely strings.